Output enhancements for `uc` command

creechy opened the pull request:
**Description of changes**

This PR addresses Issue https://github.com/unitycatalog/unitycatalog/issues/157

This change enhances the tabular output of the `uc` terminal command in a few ways
- Use the full width of the terminal for the table width
- Consider special columns like `NAME` and `ID` to be fixed with and always display their entire values
- Only fall back to equal sizes of non-special columns if any total line width is too long to fit in the terminal

The motivation is to ensure critical columns like object names are never clipped, which can make navigating entities challenging if you can't get the entire name. As an example, the sample table `marksheet_uniform` gets clipped to `marksheet..." which is not entirely useful.

Sample screenshots before this PR

<img width="1386" alt="image" src="https://github.com/unitycatalog/unitycatalog/assets/1179360/c157c1f4-1102-4540-8bbd-2680b0580fa5">

<img width="1410" alt="image" src="https://github.com/unitycatalog/unitycatalog/assets/1179360/4ccd803d-1e5c-4040-99ab-0f3f688ba312">

And after

<img width="882" alt="image" src="https://github.com/unitycatalog/unitycatalog/assets/1179360/0cf2e56d-8830-4baf-9253-d08250a3663e">

<img width="1801" alt="image" src="https://github.com/unitycatalog/unitycatalog/assets/1179360/4de70c71-a502-4f92-b32a-1f39582fddc4">

vikrantpuppala commented:
Thank you for making this change @creechy!
Could I request for a couple of additions here:
1. If you could create an issue for this change, that would be great. (If it already exists, could you please link it to this PR)
2. Could you add a few examples of what the CLI looked like without this PR vs with this PR? This would help make the improvements clear to everyone.

vikrantpuppala commented on examples/cli/src/main/java/io/unitycatalog/cli/utils/CliUtils.java:
```diff
@@ -167,35 +170,84 @@ public static void postProcessAndPrintOutput(CommandLine cmd, String output, Str
+                                    columnWidths[i] = columnWidth;
+                                }
+                            }
+                        }
```
This looks great, could you please extract out these blocks of logic into separate functions to make this more readable, I think `postProcessAndPrintOutput` has become too big

creechy commented:
@vikrantpuppala
 
> 1. If you could create an issue for this change, that would be great. (If it already exists, could you please link it to this PR)

Added an issued and put it in the PR description.

> 2. Could you add a few examples of what the CLI looked like without this PR vs with this PR? This would help make the improvements clear to everyone.

Added some before and after screen shots in the description.

vikrantpuppala commented on examples/cli/src/main/java/io/unitycatalog/cli/utils/CliUtils.java:
```diff
@@ -151,14 +147,145 @@ public static String preprocess(String value, int length) {
+
+    private static int getOutputWidth() {
+        // TODO: maybe turn this into a command line parameter.
+        String envWidth = System.getenv().get("UC_OUTPUT_WIDTH");
```
This is specifically to handle the failure in tutorial tests?

creechy replied:
Yes. I didn't look into whether you could set a "terminal width" in github actions, but I doubt it. Alternatively I could change the logic to be... if we get a non-zero width from the terminal then use it, otherwise use the original 180 character width.

vikrantpuppala replied:
I think it is fine for now, given we have a fallback default

vikrantpuppala reviewed (commented):
Looks good overall, few minor comments

creechy commented on examples/cli/src/main/java/io/unitycatalog/cli/utils/CliUtils.java:
```diff
@@ -151,14 +147,145 @@ public static String preprocess(String value, int length) {
+        return maxLineWidth;
+    }
+
+    private static void adjustColumnWidths(List<String> fields,
```
I added a little javadoc.

creechy commented on examples/cli/src/main/java/io/unitycatalog/cli/utils/CliUtils.java:
```diff
@@ -151,14 +147,145 @@ public static String preprocess(String value, int length) {
+    }
+
+    private static void processOutputAsRows(AsciiTable at, JsonNode node, int outputWidth) {
+        List<String> fields = getFields(node.get(0));
```
Done

creechy commented on examples/cli/src/main/java/io/unitycatalog/cli/utils/CliUtils.java:
```diff
@@ -151,14 +147,145 @@ public static String preprocess(String value, int length) {
+        List<String> fields = getFields(node.get(0));
+
+        int[] columnWidths = new int[fields.size()];
+        boolean[] fixedWidthColumns = new boolean[fields.size()];
```
Done

creechy commented on examples/cli/src/main/java/io/unitycatalog/cli/utils/CliUtils.java:
```diff
@@ -151,14 +147,145 @@ public static String preprocess(String value, int length) {
+            if (fixedWidthColumns[i]) {
+                fixedWidthColumnCount++;
+            }
+        }
```
Done. I implemented slightly different than the example because streams don't allow multiple terminating methods (in this case sum() and count()).

creechy commented on examples/cli/src/main/java/io/unitycatalog/cli/utils/CliUtils.java:
```diff
@@ -151,14 +147,145 @@ public static String preprocess(String value, int length) {
+            if (!fixedWidthColumns[i]) {
+                columnWidths[i] = columnWidth;
+            }
+        }
```
Done.

vikrantpuppala commented on .github/workflows/release-workflow.yml:
```diff
@@ -15,9 +15,9 @@ jobs:
+
       - name: Build tarball
-        run: sbt createTarball
+        run: build/sbt createTarball
```
why is this change showing up?

creechy replied:
> why is this change showing up?

I think its due to how merging/rebasing of main. I'll try to figure out how to fix this up.

vikrantpuppala commented:
Looks good for the most part, but I see random changes like the github workflows, tests etc showing up, can you check what's wrong?

creechy commented:
> Looks good for the most part, but I see random changes like the github workflows, tests etc showing up, can you check what's wrong?

@vikrantpuppala the other changes went away once I sync'ed the PR branch in my personal repo. Can you give it one final review now? thanks.

vikrantpuppala reviewed (approved):
lgtm, @haogang to confirm and merge, thanks!
